jwst/extract_1d/extract1d.py

```diff
@@ -1,809 +1,483 @@
-    n_srclim = len(p_src)
-    for i in range(n_srclim):
-        lower = p_src[i][0]
-        upper = p_src[i][1]
-        if independent_var.startswith("wavelength"):    # OK if 'wavelengths'
-            srclim.append([lower(lambdas), upper(lambdas)])
-        else:
-            srclim.append([lower(pixels), upper(pixels)])
-
-    if p_bkg is None:
-        nbkglim = 0
+    if profiles_2d is None:
+        profiles_2d = []
+
+    # Independent variable values for the polynomial fit.
+    y = np.linspace(-1, 1, image.shape[0])
+
+    # Build the matrix of terms that multiply the coefficients.
+    # Polynomial terms first, then source terms if those arrays
+    # are supplied.
+    coefmatrix = np.ones((image.shape[1], image.shape[0], order + 1 + len(profiles_2d)))
+    for i in range(1, order + 1):
+        coefmatrix[..., i] = coefmatrix[..., i - 1] * y[np.newaxis, :]
+
+    # Here are the source terms.
+    for i in range(len(profiles_2d)):
+        coefmatrix[..., i + order + 1] = profiles_2d[i].T
+
+    # Construct a boolean array for the pixels that are nonzero
+    # in any of our profiles.
+    pixels_used = np.zeros(image.T.shape, dtype=bool)
+    for profile_2d in profiles_2d:
+        pixels_used = pixels_used | (profile_2d.T != 0)
+    if profile_bg is not None:
+        pixels_used = pixels_used | (profile_bg.T != 0)
+    pixels_used = pixels_used & np.isfinite(image.T)
+
+    # Target vector and coefficient vector for the least squares fit.
+    targetvector = image.T.copy()
+
+    # We don't want to be ruined by NaNs in regions we are not fitting anyway.
+    targetvector[~pixels_used] = 0
+    coefmatrix_masked = coefmatrix * pixels_used[:, :, np.newaxis]
+
+    # Weighting goes here.  If we are using inverse variance weighting,
+    # weight here is the square root of the inverse variance.
+    if weights is not None:
+        coefmatrix_masked *= weights.T[:, :, np.newaxis]
+        targetvector *= weights.T
+
+    # Products of the coefficient matrices suitable for passing to
+    # linalg.solve.  These are matrices of size (npixels, npar, npar)
+    # and (npixels, npar).
+    matrix = np.einsum('lji,ljk->lik', coefmatrix_masked, coefmatrix_masked)
+    vec = np.einsum('lji,lj->li', coefmatrix_masked, targetvector)
+
+    return matrix, vec, coefmatrix, coefmatrix_masked
+
+
+def _fit_background_for_box_extraction(
+        image, profiles_2d, variance_rn, variance_phnoise, variance_flat,
+        profile_bg, weights, bg_smooth_length, bkg_fit_type, bkg_order):
+    """Fit a background level for box extraction."""
+
+    # Start by copying the input image
+    input_background = image.copy()
+
+    # Smooth the image, if desired, for computing a background.
+    # Astropy's convolve routine will replace NaNs.  The convolution
+    # is done along the dispersion direction.
+    if bg_smooth_length > 1:
+        if not bg_smooth_length % 2 == 1:
+            raise ValueError("bg_smooth_length should be an odd integer >= 1.")
```

Comment:
If the value for the background smoothing length is an even value, could we round it down to the nearest odd value and report that that has been done instead of throwing an error?  I'm not sure how common it is to use the background smoothing, but it may be preferable to have the pipeline use a sensible value rather than fail to finish running a pipeline.

If it does make sense to round to a safe value, it might make the most sense to do that earlier (e.g., in extract_1d_step) and report it just once, rather than with each loop through extract1d.

Reply:
I think this option is rarely used, but it does make sense to check it earlier.  I left this error clause as is, for any lower-level users, but added some guardrails to the get_extract_parameters function, where some of the other input parameters are conditioned.  It should now warn and round to something sensible for anything other than 0 or a positive odd integer value.